src/hera/workflows/v5/dag.py

```diff
@@ -0,0 +1,70 @@
+from __future__ import annotations
+
+from typing import Any, List, Optional
+
+from hera.workflows.models import DAGTask
+from hera.workflows.models import DAGTemplate as _ModelDAGTemplate
+from hera.workflows.models import Template as _ModelTemplate
+from hera.workflows.v5._mixins import _DAGTaskMixin, _IOMixin, _TemplateMixin
+
+
+class DAG(_IOMixin, _TemplateMixin):
+    name: str
+    fail_fast: Optional[bool] = None
+    target: Optional[str] = None
+    tasks: List[DAGTask] = []
+    _task_queue: List[_DAGTaskMixin] = []
+
+    def __enter__(self) -> DAG:
+        """Enter the context of the workflow"""
+        from hera.workflows.v5._context import _context
+
+        _context.enter(self)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        from hera.workflows.v5._context import _context
+
+        for task in self._task_queue:
+            self.tasks.append(task._build_dag_task())
+
+        _context.exit()
+
+    def _add_sub(self, node: Any):
+        self.add_task(node)
+
+    def add_task(self, task: _DAGTaskMixin) -> None:
+        self._task_queue.append(task)
+
+    def _build_template(self) -> _ModelTemplate:
```

Comment:
Can we resolve the task queue here instead of when we exit the context?

That way explicit tasks added outside of the context handler will also be accounted for.

```py
tasks = []
for task in self._task_queue:
	if isinstance(task, DAGTask):
		tasks.append(task)
	elif isinstance(task, _DagTaskMixin):
		tasks.append(task._build_dag_task())
```